### PR TITLE
Fix/27165: don't allow go to next step if empty waypoint

### PR DESCRIPTION
**Pull request by DylanDylann**
### Details

In distance request, prevent request money if empty waypoint
### Fixed Issues

$ https://github.com/Expensify/App/issues/27383
PROPOSAL: https://github.com/Expensify/App/issues/27383#issuecomment-1718148390

### Tests

1. Request money -> distance
2. Change /distance in the URL to /participants
3. Verify that the App redirect to Distance request page
- [x] Verify that no errors appear in the JS console

### Offline tests

Same above
### QA Steps

1. Request money -> distance
2. Change /distance in the URL to /participants
3. Verify that the App redirect to Distance request page
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/141406735/2128634b-047a-42fc-a1a7-3f34d38e65fd

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/141406735/198fa9e0-23a8-4e9c-babd-e60432c43280

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/141406735/c853c79e-65cc-4fe8-bd4e-2e007767acee

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/141406735/755c6462-cf79-4cfe-93cd-b74e9a132d5d

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/141406735/f2b6739a-6f2b-4240-ab2d-41ed5d297f5e

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/141406735/ac9df1d9-9d42-46de-96a7-b7a6fb84ee61

</details>

### CODE REVIEW

File: src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js
Comment (tgolen) [inline src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js:61]: Is this the same as doing `lodashHas(transaction, 'comment.waypoints.waypoint0');`? If so, can you please simplify it in this way?

Another alternative that is maybe more accurate would be:

```
const hasWaypoints = _.size(lodashGet(transaction, 'comment.waypoints')) > 2;
```

Any transaction with 1 or less waypoints should be considered, invalid.

**Review by tgolen — COMMENTED** [review]
Thank you for using withOnyx to load the transaction!

File: src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js
Comment (DylanDylann) [reply src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js]: @tgolen Thanks for your suggestion. I updated the logic for more accuracy based on the logic we used in DistanceReuquest https://github.com/Expensify/App/blob/5cf1904d08ea09c70db40151a14406dd5c1d956c/src/components/DistanceRequest.js#L293

Anyway, we need to resolve this problem https://github.com/Expensify/App/pull/25786#pullrequestreview-1647770453 first and then I can re-test this PR

**Comment by DylanDylann** [comment]
@allroundexperts @tgolen Currently, We have a general bug that relates to shouldReset logic as mentioned in https://github.com/Expensify/App/issues/26946#issuecomment-1740415582 and https://github.com/Expensify/App/issues/28566

Context: 
1. Go to Distance Request --> 
2. Fill start point and finish point --> 
3. Go to Participant Page --> 
4. Go to Confirmation Page --> 
5. Go back 2 times to Distance Request --> 
6. Click Next Button

After step 3, id field in iou will be `request${reportID of participant}`. When clicking next button at step 6, we miss reportID of previous participant 
https://github.com/Expensify/App/blob/c415f7b2758aeab8e06913a634f4079a1c12a2d8/src/libs/actions/IOU.js#L2252

It makes shouldReset is true and iou field is reset, all data is lost

**Comment by DylanDylann** [comment]
IMO, we should hold this PR until https://github.com/Expensify/App/issues/26538 is resolved

**Comment by tgolen** [comment]
Yeah, either we should hold this PR until my refactorings are done, or merge this PR with the known `shouldReset` bug. It's up to you.

**Comment by DylanDylann** [comment]
@tgolen I think we should hold this PR to avoid unexpected behavior for users. If it is ok, please help to hold it. Thanks

**Comment by tgolen** [comment]
Just an FYI that [my PR](https://github.com/Expensify/App/pull/28618) went into review yesterday.

**Comment by DylanDylann** [comment]
@allroundexperts It seems there are no blocker now, could you help to check? thanks

**Comment by DylanDylann** [comment]
Bump @allroundexperts Could you help to take a look at this issue?

**Comment by DylanDylann** [comment]
@allroundexperts It seems some other change makes this PR no longer hold. I can remove the hold label now

File: src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js
Comment (DylanDylann) [reply src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js:165]: @allroundexperts I have tried but It doesn't work for me

Comment (DylanDylann) [reply src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js:165]: We still use it in App/src/pages/EditRequestPage.js

**Comment by DylanDylann** [comment]
@allroundexperts Bump please help to take a look at this PR. Thanks

**Comment by DylanDylann** [comment]
@allroundexperts Bump could you help to check this PR?

**Comment by tgolen** [comment]
I'm not sure why this was taken off HOLD. [The PR](https://github.com/Expensify/App/pull/28618) that it was held on has not been merged.

> Yeah, either we should hold this PR until my refactorings are done, or merge this PR with the known shouldReset bug. It's up to you.

Did you decide to move ahead with merging this with the known `shouldReset` bug @DylanDylann?

**Comment by DylanDylann** [comment]
@tgolen We decided to hold this PR because of [this reason](https://github.com/Expensify/App/pull/28308#issuecomment-1744131532). But in some days ago, I re-tested this PR and saw that It works well. It seems some recent changes make this PR isn't blocked anymore.

**Comment by tgolen** [comment]
OK, as long as the `shouldReset` bug isn't an issue, I'm fine with that. Thanks for explaining!

**Comment by DylanDylann** [comment]
@allroundexperts Bump, please help to check this PR? Thanks

**Review by allroundexperts — APPROVED** [review]
Tests well!

**Review by tgolen — APPROVED** [review]

**Review by roryabraham — APPROVED** [review]

**Pull request merged**
